### ISSUE TITLE
Adafactor foreach impl performance tracker

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#133367]
### 🚀 The feature, motivation and pitch

Currently, Adafactor has both foreach and forloop implementations, but the performance of the foreach impl could be significantly improved with the following added foreach ops:

A. 
- [ ] `_foreach_mm(xs, ys)`. 
This is definitely the most challenging to implement well as it would be the first non pointwise/reduction foreach op. It is also unclear how to horizontally fuse multiple mms with different dimensions. Don't worry, the rest of the list is much more actionable.
```
var_estimates = [
                row_var @ col_var 
                for row_var, col_var in zip(device_row_vars, device_col_vars)
            ]
# becomes
var_estimates = torch._foreach_mm(device_row_vars, device_col_vars)
```

B.
- [ ] `_foreach_norm(tensors, dim=None, keepdim=False)`
We have a `_foreach_norm` today that reduces to Scalars. We can assume for the purposes of this signature that every tensor in `tensors` can be reduced along `dim`.
```
row_means = [torch.norm(grad, dim=-1, keepdim=True) for grad in device_grads]
...
col_means = [torch.norm(grad, dim=-2, keepdim=True) for grad in device_grads]
# becomes
row_means = torch._foreach_norm(device_grads, dim=-1, keepdim=True]
...
col_means = torch._foreach_norm(device_grads, dim=-2, keepdim=True]
```

C. 
- [ ] `_foreach_rsqrt(tensors)`
Self explanatory. This may be the easiest AI from this list.
```
torch._foreach_sqrt_(var_estimates)
torch._foreach_reciprocal_(var_estimates)
# becomes
torch._foreach_rsqrt_(var_estimates)
```

D. 
- [ ] `_foreach_lerp(xs: TensorList, ys: TensorList, weight: ScalarList)`
This requires adding a new overload that takes ScalarList and not just Scalar. This would be highest impact to effort value as we'd be going from 9 kernel launches to 3.
```
torch._foreach_mul_(device_row_vars, beta2_ts)
torch._foreach_mul_(row_means, one_minus_beta2_ts)
torch._foreach_add_(device_row_vars, row_means)
...
torch._foreach_mul_(device_col_vars, beta2_ts)  # type: ignore[arg-type]
torch._foreach_mul_(col_means, one_minus_beta2_ts)
torch._foreach_add_(device_col_vars, col_means)  # type: ignore[arg-type]
...
torch._foreach_mul_(device_variances, beta2_ts)
torch._foreach_mul_(grads_squared, one_minus_beta2_ts)
torch._foreach_add_(device_variances, grads_squared)
...

# becomes
torch._foreach_lerp_(device_row_vars, row_means, one_minus_beta2_ts)
...
torch._foreach_lerp_(device_col_vars, col_means, one_minus_beta2_ts)
...
torch._foreach_lerp_(device_variances, grads_squared, one_minus_beta2_ts)
``` 

E.
- [ ] `_foreach_mean(tensors, dim=None, keepdim=False)`
Similar as to B above with foreach_norm, but would require writing a new op (versus just adding a schema).
```
row_var_means = [row_var.mean(dim=-2, keepdim=True) for row_var in device_row_vars]
# becomes
row_var_means = torch._foreach_mean(device_row_vars, dim=-2, keepdim=True)
```

F. 
- [ ] `_foreach_add(xs: TensorList, ys: TensorList, alpha: ScalarList)`
Similar to D, where we wish alphas could be ScalarList instead of just Scalar. I think @ad8e mentioned wanting this in another issue.
```
torch._foreach_mul_(updates, alphas)
torch._foreach_add_(device_params, updates)
# becomes
torch._foreach_add_(device_params, updates, alphas)
```

### Alternatives

Let foreach Adafactor be slow.

Figure out how to autogenerate foreach kernels.

**Comment by ad8e** [comment]
At the moment, I convert 1D tensors to lists of Scalars when interfacing with `_foreach`:
```
torch.stack(torch._foreach_norm(params))
torch._foreach_mul_(params, torch.split(ratio, 1))
torch._foreach_mul(grad, torch.split(multiplier, 1))
torch._foreach_addcmul_(params, normalized_exp_avg, torch.split(param_step_size, 1))
```
My vague recollection is that stack() -> calcs -> split() gave a tiny speedup in my optimizer, despite the stack/split overhead, because I do a sufficient number of calculations. If `_foreach_add_` was changed to allow ScalarList, I do not know if the 1D Tensor -> ScalarList -> `_foreach_add_` pathway would be more efficient than the existing 1D Tensor -> List of 0D Tensors -> `_foreach_addcmul_` pathway.

**Comment by Chillee** [comment]
@janeyx99 We should definitely add a grouped gemm implementation into PyTorch.

**Comment by vadimkantorov** [comment]
> We should definitely add a grouped gemm implementation into PyTorch.

That would be nice! There exists a related (+ there is an existing impl of grouped pointwise convs, but maybe not all memory layouts are covered there):
- https://github.com/pytorch/pytorch/issues/115852

Regarding `_foreach_mm` - btw might be a nice shortcut for computing QKV in a single-shot for legacy code which has splitted Q, K, V in separate modules. And then maybe a "flattening" shortcut could be done to keep the modules in the model intact but make the weights actually contig in memory. And then a "foreach" method could check if they indeed are contig in memory, it could schedule a single gemm instead of 3 gemms. I think this optimization of merging horizontally gemms-consuming-the-same-input exists in torch.compile, but maybe some shortcut could be helpful/more expressive in eager mode as well...

**Comment by Triang-jyed-driung** [comment]
Any progress?
This new feature would pave the way for new optimizers such as Adam-mini, Muon and Shampoo.
`_foreach_mean`, `_foreach_mm` and a bunch of other not yet implemented.

**Comment by janeyx99** [comment]
@Triang-jyed-driung there's been no progress that I know of. We are seeking community help with this issue!

**Comment by Triang-jyed-driung** [comment]
`_foreach_matmul` would be a very useful to speed up the Muon optimizer.
A possible way to fuse `_foreach_matmul` is to group all tensors with the same shape (there are only around 10 different shapes in even the most advanced models).
My workaround of `_foreach_mean` is implemented via `_foreach_norm(ord=1)` then `_foreach_div`, which definitely works, just less elegant.

**Comment by vadimkantorov** [comment]
I also wonder if foreach-kernels could be made public / dispatched from `torch.mean` / `torch.matmul` / etc (e.g. if a special `TensorList` or a `list[Tensor]` is passed in, then foreach-kernel gets invoked).

If this is supported, then maybe a single-param code version using `torch.*` methods could also work for foreach `TensorList`, thanks to the polymorphic dispatch

**Comment by JCBrouwer** [comment]
There are some bindings for CUTLASS grouped_gemm here: https://github.com/tgale96/grouped_gemm

How does it work to move something like that into pytorch proper?

**Comment by kvnloo** [comment]
I'd like to work on item B, extending "_foreach_norm" with "dim" / "keepdim".

I checked the current surface first. The existing foreach norm is still the scalar-reduction form, while the Adafactor use case needs shape-preserving partial reductions such as the row/column reductions in this issue.

Before changing the implementation, I'd like to pin down the expected behavior against individual "torch.norm" calls and then keep the initial slice narrow:

- identical "dim" / "keepdim" across the TensorList
- CPU and CUDA correctness against the for-loop reference
- dtype, empty-dimension, and unsupported fast-path fallback coverage
- an Adafactor-shaped benchmark comparing kernel count and end-to-end foreach step time

The part I want to understand first is whether the existing foreach reduction machinery can naturally support a reduction whose outputs retain dimensions, or whether this really needs a separate multi-tensor reduction path. I don't want to add an overload until that distinction is clear.

AI assistance disclosure: I used an AI assistant to help inspect the issue/current API surface and organize the test questions above. I'll validate the implementation, tests, and measurements myself.
